// @module-federation/nextjs-mf/dist/src/plugins/container/RemoveEagerModulesFromRuntimePlugin.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const sdk_1 = require("@module-federation/sdk");
const logger_1 = __importDefault(require("../../logger"));
/**
 * This plugin removes eager modules from the runtime.
 * @class RemoveEagerModulesFromRuntimePlugin
 */
class RemoveEagerModulesFromRuntimePlugin {
    /**
     * Creates an instance of RemoveEagerModulesFromRuntimePlugin.
     * @param {Object} options - The options for the plugin.
     * @param {string} options.container - The container to remove modules from.
     * @param {boolean} options.debug - Whether to log debug information.
     */
    constructor(options) {
        this.container = options.container;
        this.debug = options.debug || false;
        this.modulesToProcess = new Set();
    }
    /**
     * Applies the plugin to the compiler.
     * @param {Compiler} compiler - The webpack compiler.
     */
    apply(compiler) {
        if (!this.container) {
            logger_1.default.warn(`RemoveEagerModulesFromRuntimePlugin container is not defined: ${this.container}`);
            return;
        }
        (0, sdk_1.bindLoggerToCompiler)(logger_1.default, compiler, 'RemoveEagerModulesFromRuntimePlugin');
        compiler.hooks.thisCompilation.tap('RemoveEagerModulesFromRuntimePlugin', (compilation) => {
            compilation.hooks.optimizeChunkModules.tap('RemoveEagerModulesFromRuntimePlugin', (chunks, modules) => {
                for (const chunk of chunks) {
                    if (chunk.hasRuntime() && chunk.name === this.container) {
                        this.processModules(compilation, chunk, modules);
                    }
                }
            });
        });
    }
    /**
     * Processes the modules in the chunk.
     * @param {Compilation} compilation - The webpack compilation.
     * @param {Chunk} chunk - The chunk to process.
     * @param {Iterable<Module>} modules - The modules in the chunk.
     */
    processModules(compilation, chunk, modules) {
        for (const module of modules) {
            if (!compilation.chunkGraph.isModuleInChunk(module, chunk)) {
                continue;
            }
            if (module.constructor.name === 'NormalModule') {
                this.modulesToProcess.add(module);
            }
        }
        this.removeModules(compilation, chunk);
    }
    /**
     * Removes the modules from the chunk.
     * @param {Compilation} compilation - The webpack compilation.
     * @param {Chunk} chunk - The chunk to remove modules from.
     */
    removeModules(compilation, chunk) {
        for (const moduleToRemove of this.modulesToProcess) {
            if (this.debug) {
                logger_1.default.info(`removing ${moduleToRemove.constructor.name}`);
            }
            if (compilation.chunkGraph.isModuleInChunk(moduleToRemove, chunk)) {
                compilation.chunkGraph.disconnectChunkAndModule(chunk, moduleToRemove);
            }
        }
    }
}
exports.default = RemoveEagerModulesFromRuntimePlugin;
